dns/route53: change accessKey to be a secret ref

euank opened the pull request:
**What this PR does / why we need it**:

It's not unusual to store the aws access and secret keys together in the
same secret under different keys. This allows conveniently using that
same secret for both.

I have not yet deployed and manually tested this change, but I plan to fairly soon. Feel free to wait on that.

```release-note
The AWS Route53 dns01 provider's Access Key must now be configured as a secret reference. If you have any Issuers or CluserIssuers configured with AWS IAM credentials, they **must** be updated to use the new `accessKeyIDSecretRef` field.
```

Note, this incidentally improves error handling for the other dns providers by refactoring secret-key-data-has-that-key checking into its own function

euank commented:
I've deployed this change to my cluster and insured that the route53 provider config, using the `accessKeyIDSecretRef`, works.

munnerz commented on docs/api-types/issuer/spec.md:
````diff
@@ -126,12 +126,17 @@ clouddns:
 ```yaml
 route53:
   accessKeyID: AKIAIOSFODNN7EXAMPLE
+  accessKeyIDRef:
````
This should be `accessKeyIDSecretRef` (from types.go below)

munnerz commented on docs/examples/acme-issuer.yaml:
```diff
@@ -36,6 +36,7 @@ spec:
       - name: route53
         route53:
           # The Route53 access key ID
+          # optionally, 'accessKeyIDRef' may be used instead to reference a secret
```
accessKeyIDSecretRef

munnerz commented on docs/api-types/issuer/spec.md:
````diff
@@ -126,12 +126,17 @@ clouddns:
     key: secret-access-key
 ```
 
+*Note*: Only one of `accessKeyID` and `accessKeyIDRef` should be specified.
````
This can be removed

munnerz commented on docs/examples/cluster-issuer.yaml:
```diff
@@ -36,6 +36,7 @@ spec:
         route53:
           # The Route53 access key ID
+          # optionally, 'accessKeyIDRef' may be used instead to reference a secret
           accessKeyID: AKIADKOU3GLWAQM8WQKJ
```
Replace with `accessKeyIDSecretRef` specification

munnerz commented on pkg/apis/certmanager/v1alpha1/types.go:
```diff
@@ -133,10 +133,11 @@ type ACMEIssuerDNS01ProviderCloudflare struct {
-	HostedZoneID    string            `json:"hostedZoneID"`
-	Region          string            `json:"region"`
+	AccessKeyID        string            `json:"accessKeyID,omitempty"`
+	AccessKeyIDRef     SecretKeySelector `json:"accessKeyIDSecretRef,omitempty"`
```
I think we want to remove `omitempty` here

euank commented:
Apologies for being slow to get back to this; I've addressed your comments.

I agree it makes sense to require the secret ref if we're okay with the API breakage.

euank commented:
/hold Will merge conflict with #243, land that one first

I'll remove the hold and do the rebase/merging once that one lands.

euank commented:
/hold

(last time I had trailing stuff after /hold, I don't know if that breaks it or not)

jonboulle commented on docs/examples/acme-issuer.yaml:
```diff
@@ -35,9 +35,10 @@ spec:
-          # The Route53 access key ID
-          accessKeyID: AKIADKOU3GLWAQM8WQKJ
-          # A secretKeyRef to a the route53 secret access key
+          # A secretKeyRef to a the AWS IAM credentials
```
to a the -> to the

(driveby, hi euank!)

Capitrium commented:
@euank Is this still on hold? Seems like #243 was superseded by #363, it'd be great to get this merged in as well.

euank commented:
Thanks for the reminder @Capitrium; I'll pick this back up

euank commented:
/hold cancel

euank commented:
/retest

prow appears to be wedged

euank commented:
Rebased again, seems to be working fine on my cluster... ping @munnerz for review since you did the last round several months ago :)

munnerz commented on pkg/apis/certmanager/v1alpha1/types.go:
```diff
@@ -179,10 +179,10 @@ type ACMEIssuerDNS01ProviderCloudflare struct {
-	SecretAccessKey SecretKeySelector `json:"secretAccessKeySecretRef"`
-	HostedZoneID    string            `json:"hostedZoneID"`
-	Region          string            `json:"region"`
+	AccessKeyIDRef     SecretKeySelector `json:"accessKeyIDSecretRef"`
```
Can we update this PR to not be a breaking change? i.e. if a user specifies `accessKeyID`, it is still used. But if a user specifies `accessKeyIDSecretRef`, it takes precedence?

munnerz replied:
In a later release we can then add validation to prevent both being set, before removing the old field altogether for 1.0.

euank replied:
Sure, I can revert it back to what I [originally had](https://github.com/jetstack/cert-manager/pull/197/files/e90e28c67ddbc8af90bd9df6a0be5828fb612c47#diff-4388b350baa92adeeb5cb67f735cb969).

I think at this point, this is blocked on making sure this is a step in the right direction via discussion in #687 though

Sinjo commented:
Hey, sorry to revive the comment thread on such an old PR, but I recently ran into a situation where I wanted to do this (read the `accessKeyID` from a secret rather than having to put it in the `Issuer` resource itself).

I figure the ship has long since sailed on replacing `accessKeyID` altogether, but I was wondering if you'd be open to a PR that lets you provide one or the other (i.e. either `accessKeyID` or `accessKeyIDSecretRef`, exclusively).

I ended up using ambient credentials, but that does mean using one AWS user for the whole cert-manager installation. Not a problem for now, but could be a limiting factor in future as our setup evolves.
